### PR TITLE
Add constexpr/noexcept

### LINKED ISSUE #999: c++ methods to be marked constexpr and noexcept as appropriate
## Feature Request

We need to do a scrub on the C++ API to further annotate functions with constexpr and noexcept. Currently only const is in use.

For example,  in RationalTime,

```cpp
     double value() const {
         return _value;
     }
```
could be 
```cpp
     constexpr double value() noexcept {
         return _value;
     }
```

### PR DESCRIPTION
**Pull request by darbyjohnston**
```Fixes #999```

I tried adding constexpr/noexcept decorators to OpenTime, if this looks reasonable I can also do a pass over OpenTimelineIO. I haven't used constexpr/noexcept much so hopefully I didn't go overboard, it certainly makes the headers look more busy.

### CODE REVIEW

**Review by meshula — COMMENTED** [review]
It's the nature of C++ code to become busy like this, it's expected. That's how you know you're Modern :p It does make sense to do it, because some compilers really do take advantage of the hinting at least partially.

What prevents the TimeRange comparisons from also becoming constexpr?

**Comment by meshula** [comment]
At the previous TSC the consensus was moving to c++14. @darbyjohnston I think we can move ahead.

**Comment by darbyjohnston** [comment]
OK, I updated CMakeLists.txt to C++14. That change along with your ```fabs()``` function allowed a number of additional functions in RationalTime/TimeRange to be marked constexpr.

I'm not sure why my changes are showing this though:
```
Submodule pybind11 updated 77 files
```
Is there a way I can revert those changes?

**Comment by ssteinbach** [comment]
We should check the docs and the README.md badges as well -- might want to split the C++14 thing into its own pr and land that first just to clarify it

**Comment by meshula** [comment]
@darbyjohnston a rebase at your end to master, and a force push to this PR should resolve the spurious commits.

Landing the CMakeLists compiler version change first makes sense, just to let the CI kick through any consequences.

**Comment by darbyjohnston** [comment]
@mehsula Thanks, that didn't seem to fix it though; it looks like those changes actually got included in one of my commits, ```5400daa51f88b71e3fac2ead9b91ed1cc792c359```, possibly from merging? I guess I could delete that commit, manually redo the changes, and then another force push?

**Comment by darbyjohnston** [comment]
I created a new PR for the C++14 change:
https://github.com/PixarAnimationStudios/OpenTimelineIO/pull/1069

The only references to C++11 I saw in the docs were about the ```any/optional``` classes, which maybe doesn't need to change. And the older C++ section:
https://opentimelineio.readthedocs.io/en/latest/cxx/older.html

**Comment by meshula** [comment]
@darbyjohnston Unfortunately the history-surgery you propose is probably the only solution.

File: CMakeLists.txt
Comment (meshula) [inline CMakeLists.txt:133]: I merged this already, so hopefully a rebase makes this part of the diff go away?

File: src/opentime/rationalTime.h
Comment (meshula) [inline src/opentime/rationalTime.h:20]: some compilers are going to reject a -1ULL, we should just use `(numeric_limits<uint64_t>::max() /2)` here.

File: src/opentimelineio/composition.h
Comment (meshula) [inline src/opentimelineio/composition.h:24]:
```suggestion
    virtual std::string composition_kind() const;
```

File: src/opentimelineio/composition.h
Comment (meshula) [inline src/opentimelineio/composition.h:26]: return by ref is scary here, but I don't want to mess with it just now. The vector is const, but the Composables can be mutated, not what the intent of the const here.

File: src/opentimelineio/effect.h
Comment (meshula) [inline src/opentimelineio/effect.h:22]:
```suggestion
    std::string effect_name() const noexcept {
        return _effect_name;
```
const ref in a return makes no guarantees that the string actually lives as long as the reference, so returning const& in general is an anti-pattern.

File: src/opentimelineio/externalReference.h
Comment (meshula) [inline src/opentimelineio/externalReference.h:21]:
```suggestion
    std::string target_url() const noexcept {
```

File: src/opentimelineio/generatorReference.h
Comment (meshula) [inline src/opentimelineio/generatorReference.h:23]:
```suggestion
    std::string generator_kind() const noexcept {
```

File: src/opentimelineio/imageSequenceReference.h
Comment (meshula) [inline src/opentimelineio/imageSequenceReference.h:34]:
```suggestion
    std::string target_url_base() const noexcept {
```

File: src/opentimelineio/imageSequenceReference.h
Comment (meshula) [inline src/opentimelineio/imageSequenceReference.h:42]:
```suggestion
    std::string name_prefix() const noexcept {
```

File: src/opentimelineio/item.h
Comment (meshula) [inline src/opentimelineio/item.h:32]:
```suggestion
    optional<TimeRange> source_range () const noexcept {
```

File: src/opentimelineio/marker.h
Comment (meshula) [inline src/opentimelineio/marker.h:36]:
```suggestion
    std::string color() const noexcept {
```

File: src/opentimelineio/marker.h
Comment (meshula) [inline src/opentimelineio/marker.h:44]:
```suggestion
    TimeRange marked_range() const noexcept {
```

File: src/opentimelineio/mediaReference.h
Comment (meshula) [inline src/opentimelineio/mediaReference.h:23]:
```suggestion
    optional<TimeRange> available_range () const noexcept {
```

File: src/opentimelineio/serializableObjectWithMetadata.h
Comment (meshula) [inline src/opentimelineio/serializableObjectWithMetadata.h:20]:
```suggestion
    std::string name() const noexcept {
```

File: src/opentimelineio/timeline.h
Comment (meshula) [inline src/opentimelineio/timeline.h:37]:
```suggestion
    optional<RationalTime> global_start_time() const noexcept {
```

File: src/opentimelineio/track.h
Comment (meshula) [inline src/opentimelineio/track.h:34]:
```suggestion
    std::string kind() const noexcept {
```

File: src/opentimelineio/unknownSchema.h
Comment (meshula) [inline src/opentimelineio/unknownSchema.h:17]:
```suggestion
    std::string original_schema_name() const noexcept {
```

**Review by meshula — CHANGES_REQUESTED** [review]
Looking almost ready! Notes

- rebase for c++14
- numeric_limits instead of -1ULL
- some const& elimination on return values, where impact is low - if you haven't used the suggestion feature before, the "merge suggestion" button updates the diff to save you work

File: src/opentime/rationalTime.h
Comment (darbyjohnston) [reply src/opentime/rationalTime.h:20]: Are the parentheses ```()``` needed?

File: src/opentimelineio/composition.h
Comment (darbyjohnston) [reply src/opentimelineio/composition.h:24]: I was planning on submitting a separate PR for const cleanup changes, though I guess it could be part of this one? There's a number of these that could be changed.

File: src/opentime/rationalTime.h
Comment (meshula) [reply src/opentime/rationalTime.h:20]: that snip was from my crusty brain matter, not documentation :)

File: src/opentimelineio/composition.h
Comment (meshula) [reply src/opentimelineio/composition.h:24]: I'm inclined to rip off the bandaid and get it done as part of constexpr just because we have to address general constness with this PR. a change to C++ version is a separate concern, so it made sense to me to split it. I don't see constness as meaningfully separable though.

File: src/opentimelineio/composition.h
Comment (darbyjohnston) [reply src/opentimelineio/composition.h:26]: Out of curiosity why not change it if we're removing the other ```const&```?

Comment (meshula) [reply src/opentimelineio/composition.h:26]: the intent of const& here is that if we get children we cannot mutate them, so we wouldn't get the intended safety by removing the const&.

 In other words, children() should really be returning something like `std::vector<Immutable<Composable>>` but we haven't got an Immutable Retainer thingy or a ConstRetainer, or a Retainer<const T>, as far as I know. So making that change seems out of scope to me for this change, as we'd first need to figure out what the returned vector should really be, possibly write a new class for it, then patch it in everywhere appropriate.

Comment (darbyjohnston) [reply src/opentimelineio/composition.h:26]: In this particular case I don't think the ```const&``` was meant to imply immutable Composables, but rather as an optimization/pessimization to avoid copying the list. For example:
https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/835fc5c12791dd29c333e5c6d7e633565b2d792f/src/opentimelineio/timeline.cpp#L35

Comment (darbyjohnston) [reply src/opentimelineio/composition.h:26]: Hmm, what about this:
https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/835fc5c12791dd29c333e5c6d7e633565b2d792f/src/opentimelineio/serializableObjectWithMetadata.h#L28

Replace with ```set_metadata```?

Comment (meshula) [reply src/opentimelineio/composition.h:26]: Since all we are doing is exposing a member with full access, it seems like the member should be public, and the accessors omitted....

Comment (meshula) [reply src/opentimelineio/composition.h:26]: I agree with your point, but not only should the children list be immutable, but c in the `auto c` would itself should be const. I think a reasonable interim would be to actually return a copy of the vector, since the children are reference counted...

Comment (darbyjohnston) [reply src/opentimelineio/composition.h:26]: I personally like the accessors since it gives us the ability to add extra functionality later, and seems similar to the rest of the API. My vote would be for ```set_metadata()```.

Comment (meshula) [reply src/opentimelineio/composition.h:26]: I think I agree? I'd prefer

```cpp
void set_metadata(const AnyDictionary&);
void merge_metadata(const AnyDictionary&);
```
These aren't accessors, but the idea that well-meaning arbitrary end users can reach in and tickle the internal state gives me the shivers. I prefer a functional approach for this kind of thing.

Comment (meshula) [reply src/opentimelineio/composition.h:26]: In order to unblock and land this work, I'd like to defer this discussion about accessors to another PR or Issue.

**Review by meshula — APPROVED** [review]
This looks ready to land. Thanks again for the scrub!

**Pull request merged**
